server/src/main/java/org/elasticsearch/indices/recovery/plan/RecoveryPlannerService.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.indices.recovery.plan;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.store.Store;
+import org.elasticsearch.index.store.StoreFileMetadata;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.common.util.CollectionUtils.concatLists;
+
+public class RecoveryPlannerService {
+    private final Logger logger = LogManager.getLogger(RecoveryPlannerService.class);
+
+    private final ShardSnapshotsService shardSnapshotsService;
+    private final ShardRecoveryPlanner shardRecoveryPlanner;
+    private final boolean snapshotRecoveriesEnabled;
+    private final String repository;
+
+    public RecoveryPlannerService(ShardSnapshotsService shardSnapshotsService,
```

Comment:
There's only really two instance of this class in use, one that doesn't do anything with snapshots for use in tests, and another that uses all the planners. I think rather than having this generalized service it'd be simpler just to have those two instances (maybe with a common one-method interface). A lot of the abstractions should sort of evaporate if we do that.

Reply:
I meant to add:

I suspect we could fold the `ShardSnapshotsService` into the nontrivial implementation, rather than having it separated as it is today. I'm not 100% sure that's the right thing to do, but this is really the only client of the `ShardSnapshotsService` so I reckon there are simplifications to be had by doing this.

Reply:
Yes, we could simplify things that way. The main reason for having `ShardSnapshotsService` as a separate class was mainly around testing, in `RecoveryPlannerServiceTests` we take advantage of that injecting custom snapshot responses for each test case. I would be inclined in keeping the `ShardSnapshotsService`.

Reply:
acked, yes I see. I think I'd just be subclassing the `RecoveryPlannerService` to override a method instead of composing these two things together, but I can see the argument for doing it this way.